archive/2025-11-01-agents-rule-of-two-a-practical-approach-to-ai-agent-security.md

```diff
@@ -0,0 +1,321 @@
+- In an **\[AC\]** configuration, the agent won’t have access to any sensitive data or systems (for instance operating in a test environment for training), so any prompt injection that reaches the agent will result in no meaningful impact.
+- In an **\[AB\]** configuration, the agent can only send new emails to trusted recipients or once a human has validated the contents of the draft message, preventing the attacker from ultimately completing their attack chain.
+
+With the Agents Rule of Two, agent developers can compare different designs and their associated tradeoffs (such as user friction or limits on capabilities) to determine which option makes the most sense for their users’ needs.
+
+## Hypothetical Examples and Implementations of the Agents Rule of Two
+
+Let’s look at three other hypothetical agent use cases to see how they might choose to satisfy the framework.
+
+**Travel Agent Assistant \[AB\]**
+
+- This is a public-facing travel assistant that can answer questions and act on a user’s behalf.
+- It needs to search the web to get up-to-date information about travel destinations \[A\] and has access to a user’s private info to enable booking and purchasing experiences \[B\].
+- To satisfy the Agents Rule of Two, we place preventative controls on its tools and communication \[C\] by:
+  - Requesting a human confirmation of any action, like making a reservation or paying a deposit
+  - Limiting web requests to URLs exclusively returned from trusted sources like not visiting URLs constructed by the agent
+
+**Web Browsing Research Assistant \[AC\]**
+
+- This agent can interact with a web browser to perform research on a user’s behalf.
+- It needs to fill out forms and send a larger number of requests to arbitrary URLs \[C\] and must process the results \[A\] to replan as needed.
+- To satisfy the Agents Rule of Two, we place preventative controls around its access to sensitive systems and private data \[B\] by:
+  - Running the browser in a restrictive sandbox without preloaded session data
+  - Limiting the agent’s access to private information (beyond the initial prompt) and informing the user of how their data might be shared
+
+**High-Velocity Internal Coder \[BC\]**
+
+- This agent can solve engineering problems by generating and executing code across an organization’s internal infrastructure.
+- To solve meaningful problems, it must have access to a subset of production systems \[B\] and have the ability to make stateful changes to these systems \[C\]. While human-in-the-loop can be a valuable defense-in-depth, developers aim to unlock operation at scale by minimizing human interventions.
+- To satisfy the Agents Rule of Two, we place preventive controls around any sources of untrustworthy data \[A\] by:
+  - Using author-lineage to filter all data sources processed within the agent’s context window
```

Comment:
Author lineage as a way to filter out untrusted data from code sounds very difficult. Most commits aren't signed anyways.